ethcore/src/engines/clique/util.rs

```diff
@@ -0,0 +1,81 @@
+use ethereum_types::{Address, H256};
+use lru_cache::LruCache;
+use parking_lot::RwLock;
+
+use engines::clique::{SIGNER_SIG_LENGTH, SIGNER_VANITY_LENGTH};
+use error::Error;
+use ethkey::{public_to_address, recover as ec_recover, Signature};
+use types::header::Header;
+
+/// How many recovered signature to cache in the memory.
+pub const CREATOR_CACHE_NUM: usize = 4096;
+lazy_static! {
+	/// key: header hash
+	/// value: creator address
+	static ref CREATOR_BY_HASH: RwLock<LruCache<H256, Address>> = RwLock::new(LruCache::new(CREATOR_CACHE_NUM));
+}
+
+/// Recover block creator from signature
+pub fn recover_creator(header: &Header) -> Result<Address, Error> {
+	// Initialization
+	let mut cache = CREATOR_BY_HASH.write();
+
+	if let Some(creator) = cache.get_mut(&header.hash()) {
+		return Ok(*creator);
+	}
+
+	let data = header.extra_data();
+	let mut sig_data = data[data.len() - SIGNER_SIG_LENGTH..].to_vec();
+	sig_data.resize(SIGNER_SIG_LENGTH, 0);
+
+	let mut sig = [0; SIGNER_SIG_LENGTH];
+	sig.copy_from_slice(&sig_data[..]);
+
+	let reduced_header = &mut header.clone();
+	reduced_header.set_extra_data(data[..data.len() - SIGNER_SIG_LENGTH].to_vec());
+
+	let msg = reduced_header.hash();
+	let pubkey = ec_recover(&Signature::from(sig), &msg)?;
+	let creator = public_to_address(&pubkey);
+
+	cache.insert(header.hash(), creator.clone());
+	Ok(creator)
+}
+
+
+/// Extract signer list from extra_data.
+///
+/// Layout of extra_data:
+/// ----
+/// VANITY: 32 bytes
+/// Signers: N * 32 bytes as hex encoded (20 characters)
+/// Signature: 65 bytes
+/// --
+pub fn extract_signers(header: &Header) -> Result<Vec<Address>, Error> {
+	let data = header.extra_data();
+
+	if data.len() <= SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH {
+		return Err(Box::new("Invalid extra_data size.").into());
+	}
+
+	// extract only the portion of extra_data which includes the signer list
+	let signers_raw = &data[(SIGNER_VANITY_LENGTH)..data.len() - (SIGNER_SIG_LENGTH)];
+
+	if signers_raw.len() % 20 != 0 {
```

Comment:
Like I mentioned elsewhere, having a variable for the 20 would be better for conveying what it represents

Reply:
done.